README.md
[![Build Status](https://travis-ci.org/happner/happn-logger.svg?branch=master)](https://travis-ci.org/happner/happn-logger)

# happn-logger

`npm install happn-logger --save`

Logger using log4js.

One instance of the logger is shared processwide. (global)

## Usage

With context and component.

```javascript
var Logger = require('happn-logger');
var context = Logger.createContext('ContextName');
var log = context.createLogger('ComponentName');

log.info('message');
// [ INFO] - 1ms    ContextName (ComponentName) message

// log can create a new component logger in the same context
var log2 = log.createLogger('AnotherComponent');

```

With component only.

```javascript
var Logger = require('happn-logger');
var log = Logger.createLogger('ComponentName');

log.info('message');
// [ INFO] - 1ms    (ComponentName) message
```

### Context

The context of the logger can be modified. The change is applied to all loggers in that context.

```javascript
var Logger = require('happn-logger');
var logContext = Logger.createContext('ContextName');
var log1 = logContext.createLogger('ComponentName1');
var log2 = logContext.createLogger('ComponentName2');

log1.info('message');
log2.info('message');
// [ INFO] - 0ms    ContextName (ComponentName1) message
// [ INFO] - 0ms    ContextName (ComponentName2) message

logContext.context = 'Renamed';
log1.info('message');
log2.info('message');
// [ INFO] - 0ms    Renamed (ComponentName1) message
// [ INFO] - 0ms    Renamed (ComponentName2) message
```

The context can be renamed from any one of the loggers and it will apply to all in the context.

```javascript
log1.context('RenamedAgain');

log1.info('message');
log2.info('message');
// [ INFO] - 0ms    RenamedAgain (ComponentName1) message
// [ INFO] - 0ms    RenamedAgain (ComponentName2) message
```

### Format string

When logging it is best not to pre-assemble the string to be logged because the cost of doing so is incurred even when the level being logged-to is not enabled.

eg. `log.debug('number: ' + theNumber);`

Instead, use a format string in the first argument. The logger will perform the string assembly internally __only if the log level is enabled__.

eg. `log.debug('number: %d, string: %s, json: %j', theNumber, theString, theObject);`

### Levels

```javascript
log.fatal('message');
log.error('message');
log.warn('message');
log.info('message');
log.debug('message');
log.trace('message');

log.log('message'); // goes to debug
```

Debug and trace messages can also be logged as follows

```javascript
log.$$DEBUG('message');
log.$$TRACE('message');
```

This allows for confident matchability to enable substituting all debug and trace calls out of production deployments. 

### Configuration

__It is important. The configure() must be called at least once.__

Shown with default values below.

```javascript
var Logger = require('happn-logger');

Logger.configure({
  logCacheSize: 50,
  logLevel: 'info',
  logTimeDelta: true,
  logStackTraces: true,
  logComponents: [],
  logMessageDelimiter: '\t',
  logDateFormat: null,
  logLayout: null,
  logFile: null,
  logFileMaxSize: 20480,
  logFileBackups: 10,
  logFileNameAbsolute: true,
  logger: null
});
```

#### logCacheSize

An array of recent log event messages is kept in `Logger.cache`. This specifies the size of that array.

Environment variable __LOG_CACHE_SIZE__ overrides.

eg. `LOG_CACHE_SIZE=5000 node myServer.js`

#### logLevel

off, fatal, error, warn, info, debug, trace, all

Environment variable __LOG_LEVEL__ overrides.

eg. `LOG_LEVEL=off node myServer.js`

#### logTimeDelta

Log the number of milliseconds since the previous log message.

#### logStackTraces

If the last argument to the logger is an error, log the stack trace in a second message.

#### logComponents

An array of log component names. Filters out all info, debug and trace messages from all component not in the list.

Environment variable __LOG_COMPONENTS__ overrides.

eg. `LOG_COMPONENTS=component1,component2 node myServer.js`

#### logMessageDelimiter

The delimiter appears between the timeDelta and the message.

#### logDateFormat

By default no date appears in the log to console. However, dates are looged when the console is redirected (piped) to file. By specifying this format dates also be written to console.

eg. `'yyyy-MM-dd hh:mm:ss.SSS'`

#### logLayout

This allows for specifying of the full log line layout. 

eg.
```javascript
  ...
  logLayout: {
    type: 'pattern',
    pattern: '%d{yyyy-MM-dd hh:mm:ss.SSS} [%[%5.5p%]] - %m'
  },
  ...
```

#### logFile

If specified - the logger logs to file and console. 

__Note: Using the file logger will in most cases fail to capture a logged fatal event that brought your server down. The file write is asynchronous and the exiting process does not wait. The log from console redirected to file will include the error because console writes are synchronous.__

#### logFileMaxSize, logFileBackups

Limit disk resources used by log file. This specifies how big the log file is allowed to get before 'archiving' it, and how many archive backups should be kept.

It creates `.log` (the current log), `.log.1` (the previous log), `.log.2` (older), etc

#### logFileLayout

Same as logLayout above - but for the logFile not the console.

#### logFileNameAbsolute

Specifies if the specified logFile is an "absolute" filename. Don't know why it matters. It's a log4js parameter.

#### logger

Enables passing in a full log4js config. All log4js config fragments from the preceding list are ignored in favour of this full config.

eg.
```javascript
  ...
  logger: {
    appenders: [
      {
        type: "console",
        layout: {
          type: 'pattern',
          pattern: '%d{yyyy-MM-dd hh:mm:ss.SSS} [%[%5.5p%]] - %m'
        }
      }
    ]
  },
  ...
```


### Events

The logger emits an event before and after each log message. This enables console applications, terminals and similar to prepare for and/or recover from the logger performing a console write.

eg.
```javascript

Logger.emitter.on('before', function() {
  // Runs immediately before every log write.
});

Logger.emitter.on('after', function() {
  // Runs immediately after every log write.
});

```



RELEASES.md
## 1.0.0 (2020-11-16)
## 1.0.0 (2020-11-19)
- SMC-1269 - JSON logging - setLevel for tests up stack
## 1.0.0 (2020-11-16)
- SMC-1269 - JSON logging
## 0.1.2 (2019-05-10)

- fix #2 - pad ms delay string to length 5 to reduce jagged margins

test/logger_instance_test.js
var should = require('chai').should();
var LoggerInstance;

describe('LoggerInstance', function() {

  beforeEach(function() {
    delete require.cache[require.resolve('../lib/logger_instance')];
    LoggerInstance = require('../lib/logger_instance');
  });

  context('log functions', function(){

    it('defines a function to emit a log message at each level', function() {
      var log = new LoggerInstance;
      log.fatal   .should.be.an.instanceof(Function);
      log.error   .should.be.an.instanceof(Function);
      log.warn    .should.be.an.instanceof(Function);
      log.info    .should.be.an.instanceof(Function);
      log.debug   .should.be.an.instanceof(Function);
      log.$$DEBUG .should.be.an.instanceof(Function);
      log.trace   .should.be.an.instanceof(Function);
      log.$$TRACE .should.be.an.instanceof(Function);
    });

    it('calls the log function', function(done) {
      var log = new LoggerInstance({
        logWriter: {
          isInfoEnabled: function() {
            return true;
          }
        },
        log: function(level, context, component, message) {
          level.should.equal('info');
          context.should.equal('context');
          component.should.equal('component');
          message.should.equal('message');
          done();
        }
      }, {value:'context'}, 'component');

      log.info('message');
    });

    it('tests of level is enabled before logging', function(done) {
      var called = false;
      var log = new LoggerInstance({
        logWriter: {
          isInfoEnabled: function() {
            return false;
          }
        },
        log: function(level, context, component, message) {
          called = true;
        }
      }, 'context', 'component');
      setTimeout(function() {
        called.should.equal(false);
        done();
      }, 10);
      log.info('message');
    });

    it('formats the log text from multiple arguments', function(done) {
      var log = new LoggerInstance({
        logWriter: {
          isInfoEnabled: function() {
            return true;
          }
        },
        log: function(level, context, component, message) {
          message.should.equal('string: STRING, number NUMBER, json {"json":"data"}');
          done();
        }
      }, 'context', 'component');

      log.info('string: %s, number %s, json %j', 'STRING', 'NUMBER', {json: 'data'});
    });

    it('includes the array of log arguments in log call if logStackTraces is enabled', function(done) {
      var log = new LoggerInstance({
        logStackTraces: true,
        logWriter: {
          isInfoEnabled: function() {
            return true;
          }
        },
        log: function(level, context, component, message, array) {
          array.should.eql([
            'string: %s, number %s, json %j',
            'STRING',
            'NUMBER',
            {json: 'data'}
          ]);
          done();
        }
      }, 'context', 'component');
      log.info('string: %s, number %s, json %j', 'STRING', 'NUMBER', {json: 'data'});
    });

  });

  it('can create anoter component logger in the same context', function() {
    var contexts = [];
    var components = [];
    var log = new LoggerInstance({
      logStackTraces: true,
      logWriter: {
        isInfoEnabled: function() {
          return true;
        }
      },
      log: function(level, context, component, message, array) {
        contexts.push(context);
        components.push(component);
      }
    }, {value: 'context'}, 'component');

    log.info('one');
    log2 = log.createLogger('component2');
    log2.info('one');

    contexts.should.eql(['context', 'context']);
    components.should.eql(['component', 'component2']);

  })

});


lib/logger_instance.js
module.exports = LoggerInstance;

var util = require('util');

function LoggerInstance(config, context, component) {
  Object.defineProperty(this, 'config', {
    get: function() {
      return config
    }
  });
  Object.defineProperty(this, 'thisContext', {value: context});
  Object.defineProperty(this, 'component', {value: component});

  Object.defineProperty(this, 'context', {
    set: function(v) {
      this.thisContext.value = v;
    },
    get: function() {
      return this.thisContext.value;
    },
    enumerable: true
  });

  this.createLogger = this.createLogger;
  this.fatal = this.fatal;
  this.error = this.error;
  this.warn = this.warn;
  this.info = this.info;
  this.debug = this.debug;
  this.$$DEBUG = this.$$DEBUG;
  this.trace = this.trace;
  this.$$TRACE = this.$$TRACE;
}

LoggerInstance.prototype.createLogger = function(component) {
  return new LoggerInstance(this.config, this.thisContext, component);
}

LoggerInstance.prototype.fatal = function() {
  if (!this.config.logWriter) return;
  if (!this.config.logWriter.isFatalEnabled()) return;
  var array, message = util.format.apply(this, arguments);
  if (this.config.logStackTraces) array = Array.prototype.slice.call(arguments);
  this.config.log('fatal', this.thisContext.value, this.component, message, array);
}

LoggerInstance.prototype.error = function() {
  if (!this.config.logWriter) return;
  if (!this.config.logWriter.isErrorEnabled()) return;
  var array, message = util.format.apply(this, arguments);
  if (this.config.logStackTraces) array = Array.prototype.slice.call(arguments);
  this.config.log('error', this.thisContext.value, this.component, message, array);
}

LoggerInstance.prototype.warn = function() {
  if (!this.config.logWriter) return;
  if (!this.config.logWriter.isWarnEnabled()) return;
  var array, message = util.format.apply(this, arguments);
  if (this.config.logStackTraces) array = Array.prototype.slice.call(arguments);
  this.config.log('warn', this.thisContext.value, this.component, message, array);
}

LoggerInstance.prototype.info = function() {
  if (!this.config.logWriter) return;
  if (!this.config.logWriter.isInfoEnabled()) return;
  var array, message = util.format.apply(this, arguments);
  if (this.config.logStackTraces) array = Array.prototype.slice.call(arguments);
  this.config.log('info', this.thisContext.value, this.component, message, array);
}

LoggerInstance.prototype.log =
LoggerInstance.prototype.debug =
LoggerInstance.prototype.$$DEBUG = function() {
  if (!this.config.logWriter) return;
  if (!this.config.logWriter.isDebugEnabled()) return;
  var array, message = util.format.apply(this, arguments);
  if (this.config.logStackTraces) array = Array.prototype.slice.call(arguments);
  this.config.log('debug', this.thisContext.value, this.component, message, array);
}

LoggerInstance.prototype.trace = 
LoggerInstance.prototype.$$TRACE = function() {
  if (!this.config.logWriter) return;
  if (!this.config.logWriter.isTraceEnabled()) return;
  var array, message = util.format.apply(this, arguments);
  if (this.config.logStackTraces) array = Array.prototype.slice.call(arguments);
  this.config.log('trace', this.thisContext.value, this.component, message, array);
}


lib/logger.js
var util = require('util');

var Cache = [];

Object.defineProperty(module.exports, 'cache', {
  get: function() {
    return Cache;
  },
  enumerable: true
});

var configured = false;

var Config = {

  // keep a cache (array) of most recent n messages
  logCacheSize: process.env.LOG_CACHE_SIZE ? parseInt(process.env.LOG_CACHE_SIZE) : 50,

  // level defaults to info (fatal, error, warn, info, debug, trace)
  logLevel: process.env.LOG_LEVEL || 'info',

  // optionally display time difference between log messages
  logTimeDelta: true,

  // optionally log stack trace if last arg to log call is error instance
  logStackTraces: true,

  // specify to only log messages from specific components (lognames)
  logComponents: process.env.LOG_COMPONENTS ? process.env.LOG_COMPONENTS.split(',') : [],

  // specify log message parts delimiter
  logMessageDelimiter: '\t',

  // log date format (eg. 'yyyy-MM-dd hh:mm:ss.SSS')
  // it uses this above format if loging or redirecting to file
  logDateFormat: null,

  // layout (eg.see below)
  logLayout: null,

  // if config.logFile is specified:
  logFile: null,
  logFileLayout: null,
  logFileMaxSize: 20480,     // limit the file size
  logFileBackups: 10,        // keep history of 10 logfiles
  logFileNameAbsolute: true, // is the log filename an absolute path?

};

Object.defineProperty(module.exports, 'configured', {
  get: function() {
    return configured;
  },
  enumerable: true
});

Object.defineProperty(module.exports, 'config', {
  get: function() {
    return Config;
  }
});

var EventEmitter = require('events').EventEmitter;

var emitter = new EventEmitter();

Object.defineProperty(module.exports, 'emitter', {
  value: emitter
});

module.exports.configure = function(config) {
  var log4js, fileAppender, previous, defaulting;
  defaulting = false;
  configured = true;

  config = config || {};

  if (typeof config.logCacheSize == 'number')    Config.logCacheSize         = process.env.LOG_CACHE_SIZE ? parseInt(process.env.LOG_CACHE_SIZE) : config.logCacheSize;
  if (config.logLevel)                           Config.logLevel             = process.env.LOG_LEVEL || config.logLevel;
  if (typeof config.logTimeDelta == 'boolean')   Config.logTimeDelta         = config.logTimeDelta;
  if (typeof config.logStackTraces == 'boolean') Config.logStackTraces       = config.logStackTraces;
  if (config.logComponents)                      Config.logComponents        = process.env.LOG_COMPONENTS ? process.env.LOG_COMPONENTS.split(',') : config.logComponents;
  if (config.logMessageDelimiter)                Config.logMessageDelimiter  = config.logMessageDelimiter;
  if (config.logDateFormat)                      Config.logDateFormat        = config.logDateFormat;
  if (config.logLayout)                          Config.logLayout            = config.logLayout;
  if (config.logFile)                            Config.logFile              = config.logFile;
  if (config.logFileLayout)                      Config.logFileLayout        = config.logFileLayout;
  if (config.logFileMaxSize)                     Config.logFileMaxSize       = config.logFileMaxSize;
  if (config.logFileBackups)                     Config.logFileBackups       = config.logFileBackups;
  if (config.logFilelogFileNameAbsolute)         Config.logFileNameAbsolute  = config.logFileNameAbsolute;
  if (config.logger)                             Config.logger               = config.logger;
  if (config.listener)                           Config.listener             = config.listener;

  // only build the log4js config if not defined
  if (!Config.logger) {

    defaulting = true;

    // create the necessary layout if date format is specified
    if (Config.logDateFormat && !Config.logLayout) {
      Config.logLayout = {
        type: 'pattern',
        pattern: '%d{' + Config.logDateFormat + '} [%5.5p] - %m'
      };
    }

    // if TTY create default format for console
    if (process.stdout.isTTY) {
      if (!Config.logLayout) {
        Config.logLayout = {
          type: 'pattern',
          pattern: '[%[%5.5p%]] - %m'
        };
      }
    }

    // if stdout is piped to file, display date, no colour
    else {
      if (!Config.logLayout) {
        Config.logLayout = {
          type: 'pattern',
          pattern: `%d{${Config.logDateFormat || 'yyyy-MM-dd hh:mm:ss.SSS'} [%5.5p] - %m`
        }
      }
    }

    // default logger includes console
    Config.logger = {
      appenders: {
        console: {  
          type: "console",
          layout: Config.logLayout 
        }
      }
    };
  }

  if (!Config.logger.categories)
    Config.logger.categories = { 
      default: { 
        appenders: Object.keys(Config.logger.appenders), 
        level: Config.logLevel 
      }
    };
  
  //we add the raw appender and category
  Config.logger.appenders.$$RAW = {
    type: "console",
    layout: { type: 'messagePassThrough' }
  }

  Config.logger.categories.$$RAW = { 
    appenders: ['$$RAW'], 
    level: Config.logLevel 
  }

  // create additional appender for logFile if specified
  if (Config.logFile && defaulting) {
    Config.logger.appenders.file = {
      "type": "file",
      "absolute": Config.logFileNameAbsolute,
      "filename": Config.logFile,
      "maxLogSize": Config.logFileMaxSize,
      "backups": Config.logFileBackups,
    };
    if (Config.logFileLayout) {
      Config.logger.appenders.file.layout = Config.logFileLayout
    } else {
      Config.logger.appenders.file.layout = {
        type: 'pattern',
        pattern: '%d{' + (Config.logDateFormat || 'yyyy-MM-dd hh:mm:ss.SSS') + '} [%5.5p] - %m'
      }
    }
    Config.logger.categories.default.appenders.push('file'); 
  }

  log4js = require('log4js');
  log4js.configure(Config.logger);

  Config.logWriter = log4js.getLogger();
  Config.rawLogWriter = log4js.getLogger('$$RAW');
  Config.logWriter.level = Config.logLevel;
  Config.rawLogWriter.level = Config.logLevel;

  previous = previous || Date.now();

  function tryStringify(obj, level, tag) {
    const toStringify = {
      timestamp: Date.now(),
      level,
      tag : tag ? tag.toString() : undefined
    };
    let data;
    try {
      if (obj instanceof Error)
        data = obj.toString();
      else if (obj == null) {
        data = "[null]";
      } else {
        JSON.stringify(obj);
        data = obj;
      }
    } catch(e) {
      data = `not stringifyable: ${obj}`;
    }
    toStringify.data = data;
    return JSON.stringify(toStringify);
  }

  Config.$$RAW = function(obj, level = 'info', tag) {
    Config.rawLogWriter[level](tryStringify(obj, level, tag));
  }

  Config.log = function(level, context, component, message, array) {

    var now, ms, string, last;

    if (Config.logComponents.length > 0) {
      if (level != 'fatal' && level != 'error' && level != 'warn') {
        if (Config.logComponents.indexOf(component) == -1) return;
      }
    }

    string = '';
    now = Date.now();

    if (Config.logTimeDelta) {
      ms = now - previous;
      string = string + ms.toString().padStart(5, ' ') + 'ms ';
      previous = now;
    } else {
      previous = now;
    }

    if (Config.logCacheSize > 0) {
      Cache.unshift({
        timestamp: now,
        timedelta: ms,
        level: level,
        context: context,
        component: component,
        message: message
      });
      while (Cache.length > Config.logCacheSize) Cache.pop();
    }

    if (context) string = string + context + ' ';
    if (component) string = string + '(' + component + ') ';
    string = string + message;

    emitter.emit('before');

    Config.logWriter[level](string);

    if (Config.listener){
      Config.listener(level, string);
    }

    if (level == 'error')
      Config.logWriter[level](array.pop());

    if (Config.logStackTraces) {
      if (last = array[array.length -1]) {
        if (last.name && last.name == 'MeshError') {
          if (last.data && last.data.stack) {
            Config.logWriter[level](last.data.stack);
          }
        }
        else if (last.stack) {
          Config.logWriter[level](last.stack);
        }
      }
    }

    emitter.emit('after', level, string, Config.logStackTraces?last:null);
  };
}

module.exports.createContext = function(context) {
  var thisContext = {
    value: context
  }

  var thisInstance = {
    createLogger: function(component, obj) {
      return module.exports.createLogger(component, obj, thisContext);
    }
  }

  Object.defineProperty(thisInstance, 'context', {
    set: function(v) {
      thisContext.value = v;
    },
    get: function() {
      return thisContext.value;
    },
    enumerable: true,
  });

  return thisInstance;
}

module.exports.createLogger = function(component, obj, thisContext, listener) {

  var thisContext = thisContext || {
    value: undefined,
  }

  var logger = obj || function() {
    logger.debug.apply(this, arguments);
  }

  if (listener)
    Config.listener = listener;

  if (!obj) {
    logger.createLogger = function(component) {
      return module.exports.createLogger(component, null, thisContext);
    }
  }

  if (!obj) {
    Object.defineProperty(logger, 'context', {
      set: function(v) {
        thisContext.value = v;
      },
      get: function() {
        return thisContext.value;
      },
      enumerable: true,
    });
  }

  logger.on = function(event, handler) {
    emitter.on(event, handler);
  }

  logger.json = {
    fatal: (obj, tag) => {
      if (!Config.logWriter) return;
      if (!Config.logWriter.isFatalEnabled()) return;
      Config.$$RAW(obj, 'fatal', tag);
    },
    error: (obj, tag) => {
      if (!Config.logWriter) return;
      if (!Config.logWriter.isErrorEnabled()) return;
      Config.$$RAW(obj, 'error', tag);
    },
    info: (obj, tag) => {
      if (!Config.logWriter) return;
      if (!Config.logWriter.isInfoEnabled()) return;
      Config.$$RAW(obj, 'info', tag);
    },
    debug: (obj, tag) => {
      if (!Config.logWriter) return;
      if (!Config.logWriter.isDebugEnabled()) return;
      Config.$$RAW(obj, 'debug', tag);
    },
    warn: (obj, tag) => {
      if (!Config.logWriter) return;
      if (!Config.logWriter.isWarnEnabled()) return;
      Config.$$RAW(obj, 'warn', tag);
    },
    trace: (obj, tag) => {
      if (!Config.logWriter) return;
      if (!Config.logWriter.isTraceEnabled()) return;
      Config.$$RAW(obj, 'trace', tag);
    }
  }

  logger.fatal = function() {
    if (!Config.logWriter) return;
    if (!Config.logWriter.isFatalEnabled()) return;
    var array, message = util.format.apply(this, arguments);
    if (Config.logStackTraces) array = Array.prototype.slice.call(arguments);
    Config.log('fatal', thisContext.value, component, message, array);
  }

  logger.error = function() {
    if (!Config.logWriter) return;
    if (!Config.logWriter.isErrorEnabled()) return;
    var array, message = util.format.apply(this, arguments);

    var location = new Error().stack.split("\n")[2];

    if (Config.logStackTraces) array = Array.prototype.slice.call(arguments);
    else array = new Array();

    array.push(location);
    Config.log('error', thisContext.value, component, message, array);
  }

  logger.warn = function() {
    if (!Config.logWriter) return;
    if (!Config.logWriter.isWarnEnabled()) return;
    var array, message = util.format.apply(this, arguments);
    if (Config.logStackTraces) array = Array.prototype.slice.call(arguments);
    Config.log('warn', thisContext.value, component, message, array);
  }

  logger.info = function() {
    if (!Config.logWriter) return;
    if (!Config.logWriter.isInfoEnabled()) return;
    var array, message = util.format.apply(this, arguments);
    if (Config.logStackTraces) array = Array.prototype.slice.call(arguments);
    Config.log('info', thisContext.value, component, message, array);
  }

  logger.debug = function() {
    if (!Config.logWriter) return;
    if (!Config.logWriter.isDebugEnabled()) return;
    var array, message = util.format.apply(this, arguments);
    if (Config.logStackTraces) array = Array.prototype.slice.call(arguments);
    Config.log('debug', thisContext.value, component, message, array);
  }

  logger.$$DEBUG = function() {
    if (!Config.logWriter) return;
    if (!Config.logWriter.isDebugEnabled()) return;
    var array, message = util.format.apply(this, arguments);
    if (Config.logStackTraces) array = Array.prototype.slice.call(arguments);
    Config.log('debug', thisContext.value, component, message, array);
  }

  logger.trace = function() {
    if (!Config.logWriter) return;
    if (!Config.logWriter.isTraceEnabled()) return;
    var array, message = util.format.apply(this, arguments);
    if (Config.logStackTraces) array = Array.prototype.slice.call(arguments);
    Config.log('trace', thisContext.value, component, message, array);
  }

  logger.$$TRACE = function() {
    if (!Config.logWriter) return;
    if (!Config.logWriter.isTraceEnabled()) return;
    var array, message = util.format.apply(this, arguments);
    if (Config.logStackTraces) array = Array.prototype.slice.call(arguments);
    Config.log('trace', thisContext.value, component, message, array);
  }

  logger.setLevel = function(level) {
    Config.logWriter.level = level;
    Config.rawLogWriter.level = level;
  }

  return logger;
}
